pkg/operation/care/garden_health.go

```diff
@@ -0,0 +1,189 @@
+// limitations under the License.
+
+package care
+
+import (
+	"context"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	v1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	operatorv1alpha1 "github.com/gardener/gardener/pkg/apis/operator/v1alpha1"
+	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
+	"github.com/gardener/gardener/pkg/component/etcd"
+	"github.com/gardener/gardener/pkg/component/gardeneraccess"
+	"github.com/gardener/gardener/pkg/component/gardensystem"
+	"github.com/gardener/gardener/pkg/component/hvpa"
+	"github.com/gardener/gardener/pkg/component/istio"
+	"github.com/gardener/gardener/pkg/component/kubecontrollermanager"
+	"github.com/gardener/gardener/pkg/component/kubestatemetrics"
+	"github.com/gardener/gardener/pkg/component/resourcemanager"
+	"github.com/gardener/gardener/pkg/component/vpa"
+	"github.com/gardener/gardener/pkg/features"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+var (
+	requiredGardenRuntimeManagedResources = sets.New(
+		etcd.Druid,
+		kubestatemetrics.ManagedResourceName,
+		gardensystem.ManagedResourceName,
+	)
+
+	requiredVirtualGardenManagedResources = sets.New(
+		resourcemanager.ManagedResourceName,
+		gardeneraccess.ManagedResourceName,
+		kubecontrollermanager.ManagedResourceName,
+	)
+)
+
+// GardenHealth contains information needed to execute health checks for garden.
+type GardenHealth struct {
+	garden          *operatorv1alpha1.Garden
+	gardenNamespace string
+	runtimeClient   client.Client
+	clock           clock.Clock
+}
+
+// NewHealthForGarden creates a new Health instance with the given parameters.
+func NewHealthForGarden(garden *operatorv1alpha1.Garden, runtimeClient client.Client, clock clock.Clock, gardenNamespace string) *GardenHealth {
+	return &GardenHealth{
+		runtimeClient:   runtimeClient,
+		garden:          garden,
+		clock:           clock,
+		gardenNamespace: gardenNamespace,
+	}
+}
+
+// CheckGarden conducts the health checks on all the given conditions.
+func (h *GardenHealth) CheckGarden(
+	ctx context.Context,
+	conditions []gardencorev1beta1.Condition,
+	thresholdMappings map[gardencorev1beta1.ConditionType]time.Duration,
+) []gardencorev1beta1.Condition {
+
+	var (
+		systemComponentsCondition        gardencorev1beta1.Condition
+		virtualGardenComponentsCondition gardencorev1beta1.Condition
+	)
+	for _, cond := range conditions {
+		switch cond.Type {
+		case operatorv1alpha1.GardenSystemComponentsHealthy:
+			systemComponentsCondition = cond
+		case operatorv1alpha1.VirtualGardenComponentsHealthy:
+			virtualGardenComponentsCondition = cond
+		}
+	}
```

Comment:
Do we want to check that we get at least, or exactly two conditions and that both condition types are set, i.e. check that none of them is empty?

Reply:
Good question 😅 
It is the same way for `seed-care` and `shoot-care` controllers. As far as I understand, it will update the conditions from `CheckGarden`, `CheckSeed`, `Check` function parameters. If the input does not include all condition type, it will still carry out all checks, but it cannot create new conditions, because the ConditionType is missing. Thus it might return empty conditions for those. I don't know the reason for this.

Reply:
As clarified offline, we can go ahead with the existing approach for now.